### PR TITLE
Add an option to expose the allowMidWordMatch as a setting

### LINKED ISSUE #53715: suggestion score - allow first character to be a weak match
(This idea was originally posted in the repo for the python extension as issue [2097](https://github.com/Microsoft/vscode-python/issues/2097), reposted here on advice of [MikhailArkhipov](https://github.com/MikhailArkhipov). This is reworded and extended with screenshots)

Intellisense completions in VSCode are only shown after typing the first letter of a method. But I don't always remember this - I often only remember some other parts of a name.

Example: If I type 
 
    import os 
    os.dir

I would like to get "mkdir" or "makedirs" as a suggestion. This works as expected in VS Community2017 (or other IDEs like pycharm):

intellisense_vs_community2017
 
In VSCode 1.25.0 (with python extension 2018.06) I only get this:

intellisense_vscode1 25

Only if I enter "os.m" I get suggestions for "mkdir" or "makedirs". It's nice that after the initial letter I can enter any string so that "os.md" shows the suggestions I want but I don't always remember the first letter of a method. I think I'm not the only one with this problem. 

This not only affects the module os but is a general problem.

I don't know how this behavior is called. maybe fuzzy completion? Please adjust the title of this thread.

### PR DESCRIPTION
**Pull request by etriebe**
This PR fixes #53715 

This won't change the default experience but will give people an option to change the fuzzy search algorithm that is used. 

How to test:
1. Create a new C# test.cs file
2. install omnisharp
3. Populate the file with this

```cs
using System;
using System.Collections.Generic;

public class FuzzyTest
{

    public FuzzyTest()
    {
        // Type here
    }
}

public class FBUserSession
{
    string test;
    string test2;
}
```
4. In constructor, type FBUserSession, UserSession and Session. The first and last will work with the setting off.
5. Turn the new setting on 
6. Redo test for step 4 and see that all those words typed will return the FBUserSession object that the user wants

### CODE REVIEW

File: src/vs/base/common/filters.ts [hunk 1]
```diff
@@ -370,7 +370,7 @@ export function matchesFuzzy2(pattern: string, word: string): IMatch[] | null {
 export function anyScore(pattern: string, lowPattern: string, patternPos: number, word: string, lowWord: string, wordPos: number): FuzzyScore {
 	const max = Math.min(13, pattern.length);
 	for (; patternPos < max; patternPos++) {
-		const result = fuzzyScore(pattern, lowPattern, patternPos, word, lowWord, wordPos, { firstMatchCanBeWeak: false, boostFullMatch: true });
+		const result = fuzzyScore(pattern, lowPattern, patternPos, word, lowWord, wordPos, { firstMatchCanBeWeak: true, boostFullMatch: true });
```
Comment (jrieken) [inline src/vs/base/common/filters.ts:373]: Unrelated change?

File: src/vs/base/common/filters.ts [hunk 2]
```diff
@@ -547,8 +547,10 @@ export namespace FuzzyScore {
 }
 
 export abstract class FuzzyScoreOptions {
-
-	static default = { boostFullMatch: true, firstMatchCanBeWeak: false };
+	static default = {
+		boostFullMatch: true,
```
Comment (jrieken) [inline src/vs/base/common/filters.ts:551]: Cosmetic only change?

**Review by jrieken — CHANGES_REQUESTED** [review]
Thanks so far. This is already looking quite good. I left a few nit comments for seemingly un-needed changes (please PR minimally). We also need to come up with a good name, IMO `firstMatchCanBeWeak` is too technical to expose it as user-facing setting

File: src/vs/base/common/filters.ts (hunk 1 above)
Comment (etriebe) [reply src/vs/base/common/filters.ts:373]: Possibly unrelated. Or maybe we should just be using the specified default here. I'll take a look!

Comment (etriebe) [reply src/vs/base/common/filters.ts]: Changed to now take advantage of options.

**Comment by etriebe** [comment]
> Thanks so far. This is already looking quite good. I left a few nit comments for seemingly un-needed changes (please PR minimally). We also need to come up with a good name, IMO `firstMatchCanBeWeak` is too technical to expose it as user-facing setting

@jrieken what about allowMidWordMatch? Because I *think* this is what this current change will do. Or we change logic to have the inverse, which is something like, forceFirstTypedCharacterToBeStartOfWord or forceMatchOnStartOfWord. I might prefer the second.

**Comment by etriebe** [comment]
@jrieken all issues should be resolved besides the name choice which I'd love your feedback on before making a choice. 

Also, as an aside, what formatter do you use for format document to work to avoid hygiene issues? I used prettier and then the whole doc changed significantly.

**Comment by etriebe** [comment]
Any chance you can take a look? @jrieken

**Comment by jrieken** [comment]
Sorry - I have been swapped recently and I am now heading out to a longer vacation. My plan is to finally pick this up in August. Thanks for being patient.

**Comment by etriebe** [comment]
> Sorry - I have been swapped recently and I am now heading out to a longer vacation. My plan is to finally pick this up in August. Thanks for being patient.

All good! Thanks for the heads up! We're going to work on a fix internally for now and will revert that once this lands here. 😀

**Comment by etriebe** [comment]
Any chance you'll be able to take a look for August? @jrieken

File: src/vs/base/common/filters.ts [hunk 3]
```diff
@@ -378,10 +378,10 @@ export function matchesFuzzy2(pattern: string, word: string): IMatch[] | null {
 	return score ? createMatches(score) : null;
 }
 
-export function anyScore(pattern: string, lowPattern: string, patternPos: number, word: string, lowWord: string, wordPos: number): FuzzyScore {
+export function anyScore(pattern: string, lowPattern: string, patternPos: number, word: string, lowWord: string, wordPos: number, options: FuzzyScoreOptions = FuzzyScoreOptions.default): FuzzyScore {
 	const max = Math.min(13, pattern.length);
 	for (; patternPos < max; patternPos++) {
-		const result = fuzzyScore(pattern, lowPattern, patternPos, word, lowWord, wordPos, { firstMatchCanBeWeak: false, boostFullMatch: true });
+		const result = fuzzyScore(pattern, lowPattern, patternPos, word, lowWord, wordPos, options);
```
Comment (jrieken) [inline src/vs/base/common/filters.ts:384]: This is likely not even needed. We use `anyScore` only to "guess" highlights for labels when suggestions use a filter-text that's different than the label-text. In those cases we do the filtering on one string but need to do the highlight on another. There is no correctness needed, it's best-effort and could have relaxed rules already before

File: src/vs/editor/common/config/editorOptions.ts [hunk 4]
```diff
@@ -4156,6 +4165,11 @@ class EditorSuggest extends BaseEditorOption<EditorOption.suggest, ISuggestOptio
 					default: true,
 					markdownDescription: nls.localize('editor.suggest.showDeprecated', "When enabled IntelliSense shows `deprecated`-suggestions.")
 				},
+				'editor.suggest.allowMidWordMatch': {
+					type: 'boolean',
+					default: false,
+					markdownDescription: nls.localize('editor.suggest.allowMidWordMatch', "When enabled suggestions allow a first characer to be matched on a non-capital letter. This will provide more search results that may not be as accurate.")
+				},
```
Comment (jrieken) [inline src/vs/editor/common/config/editorOptions.ts:4172]: It's not just about capital character but also about character after separators, like `_` or `.`. I have started a team discussion about how to talk about this and will come back with a decision

File: src/vs/base/common/filters.ts (hunk 3 above)
Comment (etriebe) [reply src/vs/base/common/filters.ts]: Good to know! Removed updates here.

**Review by jrieken — APPROVED** [review]
Thanks. I have pushed a change with the new name for this (`matchOnWordStartOnly`) and this is now ready to be merged

**Review by sandy081 — APPROVED** [review]
